### DISCUSSION TITLE
Analytics for Github Pages

### DISCUSSION THREAD

**Discussion by Artucuno** [community/community#31474] [Other Feature Feedback, Questions, & Ideas]
Hey! I'm not sure if its already a thing but Github Pages should have a graph in the traffic tab of your Github Repo with the visitors count, country of origin, etc.

Thanks.

**(13/09/2024) EDIT:**
As of now, there is still **no way** to view Github pages traffic inside of your Github repo. If you are looking for a service with built-in analytics, I recommend using [Cloudflare pages.](https://pages.cloudflare.com/)

Cloudflare Pages also supports Jekyll and has automated builds from Git, so migration from Github Pages is super simple.

**Comment by melvincwng** [comment]
@Artucuno Hey there, I believe that feature (or at least most of it) is available in GitHub already. 

You can follow the steps below:

1. Go to your code repository
2. Look for the 'Insights' tab, click on it
3. Then click on 'Traffic' tab
4. Once there, you will see a variety of metrics like visitors, how many people cloned your repo etc

Hope this answers your question 😁

**Reply by port19x** [reply]
**This does not track website traffic, but repo traffic!**

**Reply by sourabhshete99** [reply]
does it count our own view to our repo. I see on my repo there'e 1 unique visitor, does it mean that 1 me or somebody else ?

**Reply by NadimGhaznavi** [reply]
Nice!!!! I just ripped out my <script> reference to Google analytics which I couldn't get any useful information from. Frankly, I'm very, very, very  happy to not give more of my data to Google :P

**Comment by jfullstackdev** [comment]
that sounds exactly like Google Analytics, 

well you can use Google Analytics for GitHub Pages if you are really into it, I tried it before

**Reply by port19x** [reply]
It's a shame that github doesn't provide a server side solution.
Of course you can integrate google analytics, matomo, goatcounter and whatever on literally every static site host. 
That's not really analytics support.

**Reply by mishmanners** [reply]
I use Google Analytics for my GitHub Pages sites.

**Reply by Pomax** [reply]
Why would google analytics be any more acceptable than "just use bob's third party analytics and user tracker tool, since that exists"? Github, through Microsoft, has all the metrics already: we should reasonably be able to see those for our pages.

**Reply by guest271314** [reply]
Is there a way to write to a file in the repository with GitHub Actions?

**Reply by port19x** [reply]
> Is there a way to write to a file in the repository with GitHub Actions?

Yes. https://port19.xyz/tech/github-actions-monitoring/

**Reply by guest271314** [reply]
Then we *can* write the requested URLs to a file in the repository.

**Reply by port19x** [reply]
And how would that be triggered?
There is no [actions trigger](https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows) that you can hook into your webiste via a tracking pixel or sth similar.
Even if there was, it would be just as good/bad as google analytics and not server-side.

**Reply by guest271314** [reply]
Not sure yet. That is one part of what is required.

I've thought of using the GitHub API to write a comment in a Discussion or public Gist section to facilitate writing the request URL.

**Comment by rterakedis** [comment]
Hello,

@melvincwng - the Traffic stats at the location you specified appear to be directly related to repository traffic (https://docs.github.com/en/rest/metrics/traffic). In other words, when folks view the repository via github.com, stats are tracked.

What I would like to see (and what I think @Artucuno is asking), is a set of similar traffic stats as to what GitHub Pages (project.github.io) traffic is happening: 

* Popular Content - what github.io pages are being request)
* Referring Sites - where did traffic to the GitHub Page (*.github.io/pagename.html)

My primary reason for asking this is that a tool like Google Analytics is over-engineered for these basic needs.  Also, it's far less privacy-focused.   I'm simply looking for equivalent-type of stats for GitHub Pages as what is provided for Github Repos.

**UPDATE:** I did just manage to figure out that you can get _some_ of this type of reporting if you sign up for a free Cloudflare account and enable web analytics on cloudflare.   You can read about it on the [Cloudflare Blog](https://blog.cloudflare.com/secure-and-fast-github-pages-with-cloudflare/)

**Comment by acalatrava** [comment]
I agree with this proposal and I think it should be kind of mandatory. Most of us are privacy focused folks and I’d love to have, not just analytics, but direct access to logs of every request to GitHub pages, I mean server logs. We could then use any software or tool to generate analytics from them.

**Comment by joadoumie** [comment]
Just wanted to echo the sentiment in this thread that having some built-in mechanism for: 

1. Server Logs 
2. Web Traffic analytics (basic number of requests etc.)

Would be super useful!

**Comment by svnv-svsv-jm** [comment]
Oh yes please!

**Comment by ghost** [comment]
can I haz log?

![image](https://github.com/community/community/assets/961877/6310c9f1-182c-4d5a-9e00-bb7cf5a5a19f)

**Reply by Artucuno** [reply]
Real

**Comment by emmadesilva** [comment]
Would be nice to have something super simple, like the visitor counts for different pages. Similar to Popular content on the Traffic data, but for pages hosted on Pages.

**Comment by kirkone** [comment]
This would be very useful.
The integration in the Traffic page should be good for most usecases.

**Comment by hossainalmahdi** [comment]
Google analytics can be used instead

**Reply by emmadesilva** [reply]
Three problems with that:
1. It's **absolutely overkill** for most small sites
2. It will slow down page loads 
3. It will track users

Sure, if you need to know every little detail, go for GA, but if you like me just want some stats over which pages are most popular, and the visitor count, something small on the GitHub Pages server end would be ideal.

**Reply by sanxofon** [reply]
I'm sure MS already does it, so share it

**Reply by unrealsolver** [reply]
Google analytics could be opted out. Static server file access stat - can not.

**Comment by brandonbrinkley** [comment]
Having the logs would be super helpful, even if user info has to be scrubbed. I'd be very interested in knowing what URLs are referring my pages.

**Comment by davidmegginson** [comment]
Bump! Like a lot of posters, I don't use Google Analytics for ethical reasons (third-party tracking scripts are nasty), but I'd love to have basic server-side traffic info — even just pageviews/day — to know how many people are visiting.

**Reply by Skipants** [reply]
@davidmegginson There are more ethical analytics tools you can use. Check out https://dataethics.eu/tools/ under "Alternative Webstatistic Tools"

**Reply by reukiodo** [reply]
Aren't those all still client-side implementations? Server stats are basically already collected, just want exposure to view them.

**Reply by emmadesilva** [reply]
> Aren't those all still client-side implementations? Server stats are basically already collected, just want exposure to view them.

Exactly. GitHub (presumably) already collects all the information we are interested in, we just want to see it.

**Reply by dwarfered** [reply]
Agreed, though probably no financial incentive for GitHub to add such a feature. Unless they tie it to GitHub Pro or something

**Reply by emmadesilva** [reply]
> Agreed, though probably no financial incentive for GitHub to add such a feature. Unless they tie it to GitHub Pro or something

On the other hand, it should not cost that much to implement the feature, and they shouldn't lose any money by offering it. Considering that GitHub Pages is free I don't think this should be a deal breaker. I'm assuming GitHub wants people to use Pages, so it should be in their interest to have as many good features as possible.

**Comment by mvaldesdeleon** [comment]
In case any product people read this: The number 1 reason that would make me churn right now is need to have insight into my site's traffic. I've deliberately chosen to not include any cookies nor 3rd party content in my site, and would have to rely solely on server logs to have this kind of insight. I can do without it for now, but eventually that'll be my reason to host somewhere else.

**Reply by allan-stewart** [reply]
Hear, hear!

**Comment by guest271314** [comment]
Seems like we can do this using a `ServiceWorker`, Deno Deploy or Google App Engine if Microsoft doesn't want to do it.

**Reply by davidrsch** [reply]
Hello, would you mind linking to a tutorial or demo gh-page with this solution implemented?

**Reply by guest271314** [reply]
Something like

`index.html`
```
<script type="module">
  await navigator.serviceWorker.register("./sw.js", {
    scope: "./"
  });
  navigator.serviceWorker.addEventListener("controllerchange", async (e) => {
    // Do stuff when the ServiceWorker has control of the WindowClient
    fetch(location.href, {
      method: "head"
    });
  });
</script>
```

`sw.js`
```
async function log(url) {
  return fetch("https://your/deno/deploy/instance", {
    method: "query",
    body: url
  }).catch((e) => {
    // Handle error or exception
  });
}

self.addEventListener("install", (event) => {
  console.log(event.type);
  event.waitUntil(self.skipWaiting());
});

self.addEventListener("activate", async (event) => {
  console.log(event.type);
  event.waitUntil(self.clients.claim());
});

self.addEventListener("fetch", async (event) => {
  event.respondWith(fetch(event.request));
  // We're not holding up content for analytics
  log(event.request.url);
});
```

https://docs.deno.com/deploy/kv/manual/

> Opening a database
>
> In your Deno program, you can get a reference to a KV database using [Deno.openKv()](https://deno.land/api?unstable=&s=Deno.openKv). You may pass in an optional file system path to where you'd like to store your database, otherwise one will be created for you based on the current working directory of your script.

>  Watching for updates in Deno KV
>
> You can also listen for updates from Deno KV with kv.watch(), which will emit a new value or values of the key or keys you provide.
>
> ...

> Production usage
> 
> Deno KV is available for use in live applications on [Deno Deploy](https://docs.deno.com/deploy/kv/manual/on_deploy). In production, Deno KV is backed by [FoundationDB](https://www.foundationdb.org/), the open source key-value store created by Apple.

**Reply by guest271314** [reply]
Or. log \*everything\*, persistently https://github.com/guest271314/persistent-serviceworker.

**Comment by alexvgjm** [comment]
No progress on this topic yet?

All the people I know who migrates from Pages is precisely because of the lack of this feature.

Nobody talks about exposing IP addresses or anything similar, but rather something as simple as counting daily visits. I found several threads about the same. What is so problematic, dangerous or complicated about this that these requests have been ignored for more than 7 years?

I guess Microsoft wants to somehow get rid of this service. If not, I don't understand why they don't provide something so seemingly basic.

**Reply by Tedpac** [reply]
Exactly, I regretted not using Cloudflare Pages from the beginning. Maybe at some point I will migrate there.

**Comment by durack1** [comment]
Is there a supported API that provides access to statistics that extend further back in time than the last 14 days? There have been numerous requests for this over time, e.g. https://github.com/isaacs/github/issues/399 (circa ~2015), but it wasn't obvious to me that this request has been actioned? If yes, is there a link to docs that guide use of this?

**Comment by martinristow** [comment]
Јас користам гоогле аналитика

**Comment by guest271314** [comment]
Does `user/repository/graphs/traffic` suffice?

**Reply by port19x** [reply]
[No](https://github.com/orgs/community/discussions/31474#discussioncomment-5167751)

**Reply by guest271314** [reply]
Then there's (deliberately) no solution in Microsoft GitHub world.

**Comment by tsvibt** [comment]
Bump.

**Comment by silas-hw** [comment]
This would be super useful

**Comment by hstsethi** [comment]
I don't want to use Cloudflare as it's not privacy-friendly, just like Google analytics. Does anyone know any other solution?

I tried Netlify but it asked for legal ID. Vercel does work and it also has server side HTTP logs, but I was having trouble getting my site indexed  on search engines.

Edit: My site is static and only contains HTML, CSS. I prefer not to use JS either.

**Reply by guest271314** [reply]
Deno Deploy. Make a `fetch()` request on loading GH Pages, store request URL in KV storage.

**Reply by port19x** [reply]
I'm using nearlyfreespeech.net for php hosting and they have a toggle for keeping logs

**Reply by Tedpac** [reply]
If you claim that Cloudflare generically is not privacy friendly, you are most probably right, because Cloudflare has too many services to analyze internet traffic, but if you mean Cloudflare Pages, which is the closest thing to GitHub Pages, but with analytics included, could you please point me to where you got the information to claim that it is not privacy friendly?

**Comment by rterakedis** [comment]
I started looking into https://umami.is/ as a posssible analytics script.  No opinion to offer on it yet.

**Reply by Aqueuse** [reply]
I have tried it but it doesn't track anything :-/

**Reply by mtrocadomoreira** [reply]
I managed to set up [umami](https://umami.is/) with Material for Mkdocs by following [the relevant section of this post](https://dnsmichi.at/2024/02/24/gdpr-compliant-website-analytics-with-umami-opentofu-ansible-hetzner-cloud/#more-umami-integrations). I'm using Umami Cloud's free tier, which seems pretty generous in terms of features, and I'm happy with the result so far.

**Comment by mmguero** [comment]
I would love to see this as well.

**Comment by brishtibheja** [comment]
please microsoft 🥺

**Comment by guest271314** [comment]
I bet if leo asked Microsoft would hand over the logs.

**Reply by reukiodo** [reply]
who/what is leo?

**Reply by brishtibheja** [reply]
> who/what is leo?

exactly my question

**Reply by mmguero** [reply]
I assume they mean "law enforcement officers"

**Comment by Res260** [comment]
plz implement

**Comment by Sainan** [comment]
What's interesting to me is that they do clearly measure some usage data for pages, at least enough to enforce their documented bandwidth limit, so at least some way to see how close you are to hitting a limit...?

**Comment by chrissy0** [comment]
Also looking for this.

**Comment by kylto** [comment]
I think you can implement tracking using Yandex Metrica, they provide a free tracking service

**Comment by davidmegginson** [comment]
Again, the goal is to get access to the server logs (even in a heavily redacted form), not to sneak Javascript analytics spyware into people's browsers.  In any case, third-party Javascript analytics are useless for measuring direct downloads of data (i.e. not by clicking a link on a web page) for measuring the large percentage of users who block third-party analytics trackers on web pages: only GitHub has access to that information.

**Comment by stonebeaver** [comment]
Came to this thread because I wanted to peek at visitor numbers. Let's alter the viewing angle!

Why do we feel, we would have the right to get those numbers? Because other services give them to us. FB. YT. Insta. And many many more.

What's the main effect when you provide visitor metrics? They become relevant. You start to think in visitors when authoring. When someone else gets to know your blog the first question becomes "how many" instead of "what's the topic". Let alone an unavoidable percentage of users with manipulated metrics to enhance some kind of market value.

MS did never answer. MS did never provide the metrics. Until that happens I will enjoy a niche of the internet, where honest content might stay more relevant than subscription numbers.

**Reply by guest271314** [reply]
@MaykGyver If metrics don't matter all of the metrics on GitHub's home page don't matter.

> Join the world’s most widely adopted AI-powered developer platform.
>
> Millions of developers and businesses call GitHub home
>
> Work 55% faster.[Jump to footnote 1](https://github.com/#footnote-1) Increase productivity with AI-powered coding assistance, including code completion, chat, and more.
[Explore GitHub Copilot](https://github.com/features/copilot)

That "55% faster" claim is clearly suspicious.

> 7x fastervulnerability fixes with GitHub[Jump to footnote 2](https://github.com/#footnote-2)

**Reply by jeremyredhead** [reply]
> If metrics don't matter all of the metrics on GitHub's home page don't matter.

Exactly; they don't.

**Comment by guest271314** [comment]
Couldn't Microsoft Clarity https://clarity.microsoft.com/ be integrated into GitHub?

**Reply by stonebeaver** [reply]
> Couldn't Microsoft Clarity https://clarity.microsoft.com/ be integrated into GitHub? 

@guest271314 You're lucky! [It works.](https://maykgyver.github.io/2025/02/02/ms-clarity-with-github-pages.html)

Keep in mind, that MS Clarity ist client-side tracking. So if the javascript function ist not executed by the client for any reason, the tracking won't hit. A server-side tracking would count all requests instead.

**Reply by guest271314** [reply]
Seems to me Microsoft could "integrate" their own products into GitHub.

**Reply by rterakedis** [reply]
As an FYI, there's a lot of fine print in the [Clarity FAQs](https://learn.microsoft.com/en-us/clarity/faq), particularly in the Privacy section.

**Reply by guest271314** [reply]
@rterakedis 

> Responsible AI

Ha, ha, ha. Too funny. Same spiel StackExchange tried to sell [Our partnership with Google and commitment to socially responsible AI](https://meta.stackexchange.com/questions/398127/our-partnership-with-google-and-commitment-to-socially-responsible-ai)(https://meta.stackexchange.com/questions/398127/our-partnership-with-google-and-commitment-to-socially-responsible-ai/398182#398182).

What "privacy"? There is such thing for any signal communications. Since at least ThinThread was publicly disclosed [A Good American](https://agoodamerican.org/). Really since "the burglars" exposed COINTELPRO [The Burglary That Exposed COINTELPRO](https://billmoyers.com/story/the-burglary-that-exposed-cointelpro/).

> It was pretty clear that we were building the most powerful analysis tool
that had been developed in history to monitor basically the entire world.

- Bill Binney, A Good American


You're dealing with Microsoft.With the analytics and with GitHub. 

- [Microsoft handed the NSA access to encrypted messages](https://www.theguardian.com/world/2013/jul/11/microsoft-nsa-collaboration-user-data)
- [Everything you need to know about PRISM](https://www.theverge.com/2013/7/17/4517480/nsa-spying-prism-surveillance-cheat-sheet)
- 
And then named an emulator the same name!

- [Prism](https://learn.microsoft.com/en-us/windows/arm/apps-on-arm-x86-emulation#prism)

**Comment by Sup2point0** [comment]
still waiting 👀

**Comment by guest271314** [comment]
Hmm. Create a gist. Add an initial comment containing `0`. On each request to GitHub pages use GitHub REST API to update https://docs.github.com/en/rest/gists/comments?apiVersion=2022-11-28#update-a-gist-comment the comment existing comment as number +1?

**Comment by thekangaroofactory** [comment]
Definitely useful to help prioritize our projects, and how to improve them:

A few examples how those metrics would help:
- package documentation page: from that you can focus on providing robust function documentation & support for your users
- blog page: identify most popular topics, dive into them and share more insight to readers

**Comment by thecosmicskye** [comment]
Bump? Is there an alternative free hosting site? Cloudflare Pages?

**Reply by rterakedis** [reply]
Cloudflare does offer basic analytics with Pages, though iirc it’s not enabled by default.  Umami is another free alternative.  Both seem to be privacy focused for end users.

**Comment by guest271314** [comment]
Using Deno Deploy, where it's possible to sign in with GitHub account, and export the code to GitHub.

Server
```
Deno.serve(() => new Response("", {
  cache: "no-store",
  headers: {
    "access-control-allow-origin": "https://github.com",
    "vary": "origin",
    "pragma": "no-cache",
    "cache-control": "no-cache",
  }
}));
```

Request from GitHub origin

```
fetch("https://xxx-zzz-123.deno.dev", {cache: "no-cache"})
  .then((response) => {
    console.log(response.ok);
  }).catch(console.warn);
```

Analytics on https://dash.deno.com/projects/xxx-zzz-123/analytics/24h
![deno-deploy-analytics](https://github.com/user-attachments/assets/11c6d51c-0249-47d1-aaee-e192123a909b)

**Comment by Taormina** [comment]
Github, Microsoft, this is still basic functionality that everyone wants, and you have to already be tracking on your end. Can we get this?

**Reply by guest271314** [reply]
I think Microsoft hears, but ain't listening. Deno Deploy and Vercel both have "free" tiers, and both provide analytics. In GH Pages you can make a request to either to get statistics for Gh Pages requests.

**Comment by collapsinghierarchy** [comment]
Would an e2ee solution work for you people? Currently trying to build one and wanted to get a heads-up if it might serve exactly this purpose. 

You create a keypair in the browser, add the analytics script (in the same way as it happens with GA) to your static webpage. If now a visitor is opening your page the same stats are gathered, but they are encrypted in-Browser before being streamed to some remote back-end. You can then open the dashboard, pull the encrypted analytics and decrypt them on your local machine and the JS shows the analysis results.  It's not ready yet, but your answers might provide valueable feedback!

EDIT: Now i saw that most of you want only to see the logs. That's fine. Can't solve this problem. But an e2ee would at least keep the analytics part private to you. And you don't have to track anyone. Simple metrics should already suffice like the unique visitors / acitvity duration et al.. There is no tracking required for these stats.

**Reply by guest271314** [reply]
There's options. I would just use a free tier service like Deno Deploy or Vercel to get how many times the script loads.

I think folks are kinda hopin' Microsoft will chime in and say something like "We hear ya", and just slap on one of their existing analytics products to GitHub Pages.

That has not happened.

**Reply by Taormina** [reply]
There are a million analytics solutions, it's just wild that Microsoft and GitHub have several in-house solutions to this, AND that they have to be doing this tracking internally anyway, but it's not exposed anywhere. I can grab GA, I just don't really want to have to for what should've existed for literal years at this point.

**Reply by guest271314** [reply]
GitHub *is* Microsoft. Microsoft owns GitHub  last time I checked. So, they know and have been silent on the feature request.

**Reply by Taormina** [reply]
I mean, yes, Microsoft bought GitHub, but massive corporations don't instantly merge and seamlessly operate in lock step. The whole point of posting here is to create volume and noise and engagement for whatever product manager is deciding what they should work on next.

**Reply by guest271314** [reply]
Microsoft bought GitHub. That's not a merger. That's an acquisition. 

There's plenty of volume here. GitHub management is clearly avoiding and ignoring this feature request. Not for lack of technical capability.

**Comment by GReturn** [comment]
Bump. I would love this being implemented. It's such a basic thing to have.

**Comment by NadimGhaznavi** [comment]
Why not some Javascript? You don't have to be "sneaky". You can still respect privacy while collecting the data. You could save to CSV and render using ApexCharts and do it all within your repo.

**Reply by guest271314** [reply]
What's your steps for doing that? I found that using JavaScript by way of Deno Deploy in a GitHub repository would work. What I read is that folks kinda want Microsoft to step up and provide the general mechanics in the GitHub UI and/or automation process. I'd bet that Microsoft does those analytics anyway, already.

**Reply by NadimGhaznavi** [reply]
So the flow is:

Visitor loads page
   ↓
JS sends repository_dispatch to GitHub API (no auth needed for public repos)
   ↓
GitHub Action triggers
   ↓
Action appends entry to metrics.csv
   ↓
Commit & push


All authentication is handled inside GitHub by the Actions runner using the automatic ${{ secrets.GITHUB_TOKEN }}.
Your visitors never see or handle any tokens.

**Reply by NadimGhaznavi** [reply]
Sure, it would be great if Microsoft engineered a solution, but I wouldn't hold my breath. The fact that GitHub pages is free is already pretty amazing. Also, you're **only** collecting page counts. That's the rub, you don't get "by country" metrics. You **can** get this info:

```
Data	            Example	          How to Get It
Local timezone	    America/New_York  Intl.DateTimeFormat().resolvedOptions().timeZone
Browser language	"en-US"	          navigator.language
```

So you can build pseudo location info based on timezone and the language setting the user is using....

**Reply by NadimGhaznavi** [reply]
FYI: Toronto, Canada is in the same timezone as Quito, Ecuador!

**Reply by guest271314** [reply]
I'm not a GitHub Actions expert. I'm also not interested in specifics about visit metrics, just count, if that. Can you share the complete code with instructions here?

**Reply by NadimGhaznavi** [reply]
It's just an idea, I dont' have any code.

**Comment by DeCEll-1** [comment]
Could this be implemented already?

**Reply by ross-spencer** [reply]
+1

I also think the OP's original suggestion of a graph in the repository traffic section would be good enough, and clearly only needs to be available to repos making use of GH pages.

**Comment by BfdCampos** [comment]
Mandatory monthly bump to this thread. Anything yet?

**Comment by nurzhansabitov** [comment]
Up

**Comment by deanalvero** [comment]
How is this not implemented yet? It is almost 2026!

**Comment by NadimGhaznavi** [comment]
LOL... Okay, if nobody picks it up by 2026, I'll spend some time in January doing what I can i.e. basic page count + timezone + browser language metrics.

**Reply by NadimGhaznavi** [reply]
JS is *not* my thing, but I can hack anything, some things just take longer ;)

**Reply by Taormina** [reply]
Do you work at Microsoft on Github? I don't think anyone wants a random analytics library thrown together to be added to our websites. We have hundreds of those in the wild.

We're here questioning why Github Pages still lacks the most obvious of features, that they obviously have already internally.

**Reply by reukiodo** [reply]
This is more about getting a view on the SERVER-side metrics, similar to Apache logs, not another client implementation

**Reply by guest271314** [reply]
@Taormina 

> We're here questioning why Github Pages still lacks the most obvious of features, that they obviously have already internally.

Well, because Microsoft doesn't want to. Therefore the solution as I see it is to roll your own.

**Reply by NadimGhaznavi** [reply]
No, I don't work for The Borg. I'm a Linux guy. So, good luck with that. As far as it being "a feature that's available internally", it's a bit more complicated than that. Of course MS has access to the server side logs, but GitHub is a distributed platform, so you have aggregation. There are privacy issues, so you have scrubbing. Sure it's not a BIG feature, but it's still another feature for GitHub to support. 

Working to engineer an analytics that consolidate:

- Viewcounts
- Timezone
- Browser language

-strikes me as a pretty decent workaround i.e. versus asking for a free handout on what's already a pretty amazing free platform... But, that's just my $0.02.

**Reply by NadimGhaznavi** [reply]
A more productive approach would be to petition MS for the cost of providing this feature and to fundraise to pay for it.

**Reply by guest271314** [reply]
@NadimGhaznavi I think the analytics part is doable within GitHub using Deno Deploy, as I describe above.

> A more productive approach would be to petition MS for the cost of providing this feature and to fundraise to pay for it.

Microsoft doesn't need money. It's sheer lack of will to provide this feature.

**Reply by NadimGhaznavi** [reply]
It's not "lack of will", it's an unwillingness to give away more of their resources for free. The feature doesn't exist. Creating, updating as required, and supporting it all require resources. This isn't a cookie in a jar waiting for MS to open the lid.

**Reply by guest271314** [reply]
Well, that's your opinion. Last stories I heard Bill Gates was playing 100K/200K blinds in poker. 

The capability clearly exists. In Microsoft world. They've got Web developer tools just like Alphabet.

I'm not giving Microsoft a dime. Havn't wun Windows in years and don't plan to, again. I'm typing on Debian 13 live session from a USB with Windows on the hard drive, that ain't gonna get used.

I ain't beggin' for this, either. Much more expedient to roll my own. If this was something other than a modicum of a challenge for me, and I really wanted to do this, I'd just use Deno Deploy.

**Reply by mmguero** [reply]
I highly recommend [Umami](https://github.com/umami-software/umami). Open source, selh-hosted so you're in control of the data. It's solved this problem for me. I have my Umami url and site id stored as repository secrets that get injected into the html during my deploy workflow. It works great and had all the analytics I could ask for.

**Reply by Taormina** [reply]
> @Taormina
> 
> > We're here questioning why Github Pages still lacks the most obvious of features, that they obviously have already internally.
> 
> Well, because Microsoft doesn't want to. Therefore the solution as I see it is to roll your own.

Well, Github didn't want to. THEN Microsoft bought them. They had been sleeping on this for years. But Github Pages is a feature we actively pay for. So they aren't giving anything away for free, we're questioning why what we've paid isn't enough.

Obviously the solution for anyone that wants a solution today is to roll your own, and tons of client-side analytics implementations have existed for a long time. No one is contesting that detail.

**Reply by guest271314** [reply]
Who pays for GitHub Pages?

**Reply by guest271314** [reply]
> So they aren't giving anything away for free, we're questioning why what we've paid isn't enough.

I guess that's one way of looking at it.

Another way to look at it is Microsoft _could_ be paying developers to have their gear here, instead of, for example GitLab. 

Is your strategy to keep asking until somebody in Microsoft world says, alright, for 99.99 a month, you can get IP addresses of folks who visit your GH Pages, all sorted and analyzed by Copilot, with recommendations about how to improve traffic?

**Reply by Taormina** [reply]
It's the only feature I pay for. Everything else I need is Free Tier.

For $100/month, folks won't probably bother using it. Especially since Github currently charges by the seat. We're not even saying we want IP address, and the sort of folks who avoid GA aren't the people who want Copilot recommendations. We're saying we want the absolute basics. How many times did the server serve my page? Basic server-side analytics would be plenty.

**Reply by guest271314** [reply]
Yeah, I ain'y givin' GitHub (Microsoft) a dime for analytics I can make happen myself for free.

**Reply by Taormina** [reply]
It's 2026 now!

> I initially thought that this thread was about providing a solution within the bounds of what is currently available on GitHub.

This is all we ever wanted.

**Reply by NadimGhaznavi** [reply]
Check out [GoatCounter](goatcounter.com). I copied the `script` code into a `_includes/head/custom.html` now I have a nice dashboard showing page analytics for every page visit in my custom *osoyalce.com* domain... all hosted on GitHub pages (except for the GoatCounter).

<img width="1388" height="1200" alt="Screenshot From 2026-08-22 15-18-55" src="https://github.com/user-attachments/assets/8953f667-2778-4b52-bfbd-c80d1520aca9" />

**Comment by jdmarshall** [comment]
@NadimGhaznavi What were you proposing to do about this issue?

**Reply by NadimGhaznavi** [reply]
I initially thought that this thread was about providing a solution within the bounds of what is currently available on GitHub. However, it seems this thread is focused on pressuring Microsoft to make the (scrubbed) server logs available. Given that focus, I have nothing to offer: I can't apply any pressure on The Borg :P

**Reply by jdmarshall** [reply]
I don’t think that’s consensus. I’d like to have unique visitors and page counts. The logs thing smells more like bargaining to me.

**Reply by guest271314** [reply]
I don't think there's a difference. There's no such thing as "scrubbed" server logs. If the feds ask Microsoft to turn over data, they'll turn it over. Users ask for the same data Microsoft remains silent.

**Comment by CodyCBakerPhD** [comment]
I would love this feature :)

Please GitHub, make it so!

**Comment by techtanic** [comment]
This is really a basic feature that just needs some API and UI changes to display the data that is almost surely already collected internally, we don't even need complex stats just the basic page metrics like; visitor count, that way we developers can know what's popular and focus on improving it.

Also, it helps as a motivator seeing the popularity of a site.

This is really causing me to consider moving to Cloudflare Pages

**Comment by AqRec** [comment]
lmao it's 2026 now GitHub why not just ask your Copilot to achieve this feature?

**Reply by guest271314** [reply]
It's not a lack of know-how or technology...

**Comment by guest271314** [comment]
Check this out https://github.com/varadTheDeveloper/blurit Web site https://www.blurit.org/

**Comment by ojensen5115** [comment]
I do not expect this to move the needle at all, but the lack of nginx-style logs for my personal website is what made me leave github pages. I still have the repo in github, I just have a github action that builds it and rsyncs it over to a vps I had lying around. It was actually really easy to put together, and now i'm no longer limited by the Jekyll plugins github supports. It also has the side benefit that I can take the repo private now if I want to.

**Comment by shashanks2223** [comment]
This would actually be pretty useful for personal portfolios hosted on GitHub Pages. It would be nice to have some basic analytics without having to set up another third-party service just to see things like visitors and page views.

**Reply by jdmarshall** [reply]
And which pages are getting viewed...

**Comment by lukasMega** [comment]
There should be built-in analytics provided by GitHub for GitHub Pages. Until then, we need an alternative.

I created a custom analytics tool—now anyone can create or vibe-code one. For my purposes, I built a simple tool using Deno KV as the backend (free tier). It has very basic functionality. See: https://lukasmega.github.io/deno-kv-analytics/quickstart

//update:
someone clicked, first visitor, I see this: <details><summary> click to see sceenshot 🖼️</summary>

<img width="1868" height="1839" alt="Screen Shot 2026-09-11 at 19 18 29-fullpage" src="https://github.com/user-attachments/assets/4f623596-a67a-46a9-b872-54ef167e1af4" />


</details>

**Reply by ross-spencer** [reply]
_> now anyone can create or vibe-code one_

correct.

yet they still try and charge at different tiers...

**Reply by guest271314** [reply]
Somebody marked this as "off-topic" https://github.com/orgs/community/discussions/31474#discussioncomment-17668975.

I've been using Deno's free server for a while now.
